rotkehlchen/chain/ethereum/names.py

```diff
@@ -67,56 +67,135 @@ def search_for_addresses_names(
     """
-    db_addressbook = DBAddressbook(database)
-    db_ens = DBEns(database)
+
+    prioritizer = NamePrioritizer(database)
+    prioritizer.add_fetchers({
+        'blockchain_account': blockchain_address_to_name,
+        'global_addressbook': global_addressbook_address_to_name,
+        'private_addressbook': private_addressbook_address_to_name,
+        'ethereum_tokens': token_mappings_address_to_name,
+        'hardcoded_mappings': hardcoded_address_to_name,
+        'ens_names': ens_address_to_name,
+    })
 
     with database.conn.read_ctx() as cursor:
-        reverse_ens = db_ens.get_reverse_ens(cursor, addresses=addresses).values()
-        mapped_tokens = GlobalDBHandler().get_tokens_mappings(addresses=addresses)
-        hardcoded_mappings = {}
+        settings = database.get_settings(cursor)
+
+    prioritized_addresses = prioritizer.get_prioritized_names(
+        settings.address_name_priority,
+        addresses,
+    )
+
+    return prioritized_addresses
+
+
+FetcherFunc = Callable[[DBHandler, ChecksumEvmAddress], str]
+
+
+class NamePrioritizer:
+    _fetchers: Dict[AddressNameSource, FetcherFunc] = {}
+    _db: DBHandler
+
+    def __init__(self, database: DBHandler):
+        self._db = database
+
+    def add_fetchers(self, fetchers: Dict[AddressNameSource, FetcherFunc]) -> None:
+        self._fetchers.update(fetchers)
+
+    def get_prioritized_names(
+            self,
+            prioritized_name_source: List[AddressNameSource],
+            addresses: List[ChecksumEvmAddress],
+    ) -> Dict[ChecksumEvmAddress, str]:
+        """
+        Gets the name from the name source with the highest priority.
+        Name source ids with lower index have a higher priority.
+        """
+        top_prio_names: Dict[ChecksumEvmAddress, str] = {}
+
         for address in addresses:
-            constant_name = ETHADDRESS_TO_KNOWN_NAME.get(address)
-            if constant_name is not None:
-                hardcoded_mappings[address] = constant_name
-
-        with db_addressbook.read_ctx(AddressbookType.GLOBAL) as global_cursor:
-            global_addressbook = db_addressbook.get_addressbook_entries(
-                cursor=global_cursor,
-                addresses=addresses,
-            )
-        private_addressbook = db_addressbook.get_addressbook_entries(
+            for name_source in prioritized_name_source:
+                fetcher = self._fetchers.get(name_source)
+                if not fetcher:
+                    raise NotImplementedError(
+                        f"address name fetcher for '{name_source}' is not implemented")
+
+                name = fetcher(self._db, address)
+                if not name:
+                    continue
+                top_prio_names[address] = name
+                break
+
+        return top_prio_names
+
+
+def blockchain_address_to_name(db: DBHandler, address: ChecksumEvmAddress) -> str:
+    with db.conn.read_ctx() as cursor:
```

Comment:
1. Please add a docstring

2. That's not a good approach. What you are doing here is you query all the account data and its labels from the DB. Then you iterate them and if you find the `address` argument then you return it. Otherwise you return empty string.
What you should do instead is do a direct query in the DB for this address and get the label if existing. IF not return `None`.